### PR TITLE
Update book structure and publishing

### PR DESCRIPTION
**Pull request by lholmes8**
## Description
Update book structure and publishing for efficiency and wdb save down fix

## Checklist
Before creating a pull request please make sure you have the following:

- [x] A short but descriptive title
- [x] Assignees responsible for moving the PR forward.
- [x] Colour-coded labels for categorisation.
- [x] A project to track PRs on a higher level

### CODE REVIEW

File: code/processes/bitfeed.q [hunk 1]
```diff
@@ -24,10 +24,31 @@ while[.bfx.notpconnected[];
...
 
 .bfx.publish:{[t;x].bfx.tp(`.u.upd;t;flip x)};                                                         / Publish to tickerplant after replaylog
-.bfx.checkws:{@[{x .j.j `event`cid!(`ping;1234)};;{[x].bfx.init[]}] each .bfx`quote`trade};            / Send ping event which returns pong
+.bfx.publishbook:{[price;chanID;tbl]
+    if[price in .bfx.cdepth sublist $[`bbook=tbl;desc;asc] exec price from value tbl where chanID=chanID;  / If record in top .bfx.cdepth changes publish  
+       bidtemp:0!.bfx.cdepth sublist `price xdesc select from .bfx.bbook where chanID=chanID;              / Bid book for sym 
+       asktemp:0!.bfx.cdepth sublist `price xasc select from .bfx.abook where chanID=chanID;               / Ask book for sym
+       .bfx.tp (`.u.upd;                                                                               / Publish whole book
+                 `book_bfx;
+                 (enlist [.bfx.bookchannels chanID],bidtemp[`price`noorders`tot],abs asktemp`price`noorders`tot))
```
Comment (claire-sloan) [inline code/processes/bitfeed.q]: do you need the [] here for bidtemp?

File: code/processes/bitfeed.q [hunk 2]
```diff
@@ -24,15 +24,28 @@ while[.bfx.notpconnected[];
 .bfx.tp:.servers.gethandlebytype[.bfx.tickerplanttypes;`any];
 
 .bfx.publish:{[t;x].bfx.tp(`.u.upd;t;flip x)};                                                         / Publish to tickerplant after replaylog
+.bfx.publishbook:{[price;chanID;tbl]
+    if[price in .bfx.cdepth sublist $[`bbook=tbl;desc;asc] exec price from value tbl where chanID=chanID;  / If record in top .bfx.cdepth changes publish  
+       bidtemp:0!.bfx.cdepth sublist `price xdesc select from .bfx.bbook where chanID=chanID;              / Bid book for sym 
+       asktemp:0!.bfx.cdepth sublist `price xasc select from .bfx.abook where chanID=chanID;               / Ask book for sym
+       .bfx.tp (`.u.upd;                                                                               / Publish whole book
```
Comment (claire-sloan) [inline code/processes/bitfeed.q]: align comment

File: code/processes/bitfeed.q [hunk 3]
```diff
@@ -60,20 +73,18 @@ if[not `depth in key .proc.params;
...
-      .[`.bfx.book;;_;a[1;0]] a[0],`ask`bid -1 1f ? a[1;2];                                            / If no.orders=0 remove records with 0 orders
-      {.[`.bfx.book;(x 0;y);upsert;x 1]}[a] `ask`bid a[1;2]>0
-  ];
-   if[not 99h~type a;
-       .bfx.publish[`book_bfx] enlist .bfx.bookchannels[a 0],0!'.bfx.cdepth#'((xasc;xdesc)@'`price)@'.bfx.book[a 0]`ask`bid]
-
- };
+      {[a;x]delete from x where price=a[1;0]}[a]'[`.bfx.bbook`.bfx.abook];                             / If no.orders=0 remove records with 0 orders
+      .bfx.publishbook[a[1;0];
+                       a[0];
```
Comment (claire-sloan) [inline code/processes/bitfeed.q]: can drop [] here
add comments here

**Comment by lholmes8** [comment]
There have been two main changes. Firstly the structure of the book has been changed, both in how we store the bid and ask books in the feedhandler and the schema of the book in the rdb. They are all now tables or keyed tables rather than the complex dictionary containing keyed tables which the WDB didn't like saving down. Secondly the publishing has changed. Previously a record was published any time the top 25/100 changed (depending on subscribed depth) whereas now a record of the book is only published when a price within the user specified depth changes. This means less records being published and then stored in the RDB, so should reduce memory and cpu usage by quite a bit.

File: code/processes/bitfeed.q [hunk 4]
```diff
@@ -24,15 +24,28 @@ while[.bfx.notpconnected[];
...
+.bfx.data:();                                                                                          / Create empty list to insert data into
 .bfx.book:()!();
+.bfx.bbook:([chanID:();price:()]noorders:();tot:());
+.bfx.abook:([chanID:();price:()]noorders:();tot:());
+.bfx.tbook:([]chanID:();price:();noorders:();tot:());
 .bfx.lasthb:0Np;
 .bfx.depth:(25;100) 25<.bfx.cdepth:first"I"$.proc.params`depth;                                        / Set .bfx.depth based on value of depth param
 if[not `depth in key .proc.params;
-   .bfx.cdepth:25;
+   .bfx.cdepth:12;
```
Comment (claire-sloan) [inline code/processes/bitfeed.q]: hard-coded? can we not specify what depth we want

Comment (lholmes8) [reply code/processes/bitfeed.q]: Can be input as a command line parameter but this is the default if one isn't provided

Comment (claire-sloan) [reply code/processes/bitfeed.q]: hmm any reason for it being set to 12 specifically? why was it 25 before

Comment (lholmes8) [reply code/processes/bitfeed.q]: We just wanted to reduce the number of records being published, no reason for 12 specifically

**Pull request merged**